### ISSUE TITLE
afterTest hook does not execute for tests that fail with timeOut errors

### ISSUE THREAD

**Issue by Harshad3009** [webdriverio/webdriverio#5545]
**Environment (please complete the following information):**
 - **WebdriverIO version:**  6.1.20
 - **Mode:** WDIO Testrunner
 - **If WDIO Testrunner, running sync/async:** sync
 - **Node.js version:** 12.18.1
 - **NPM version:** 6.14.5
 - **Browser name and version:** Chrome 83
 - **Platform name and version:** Ubuntu 18.04.4
 - **Additional wdio packages used (if applicable):** @wdio/spec reporter, @wdio/devtools-service, @wdio/mocha-framework, 

**Config of WebdriverIO**
```javascript
const debug = process.env.DEBUG
const defaultTimeout = 60000
const defaultCapabilities = [{ maxInstances: 5, browserName: 'chrome' }]
var slack = require('wdio-slack-service')
const video = require('wdio-video-reporter');
let tests = 0;

exports.config = {
    runner: 'local',
    specs: [
        './test/specs/**/*.js'
    ],
    maxInstances: debug ? 1 : 10,
    capabilities: debug ? [{ browserName: 'chrome' }] : defaultCapabilities,
    execArgv: debug ? ['--inspect'] : [],
    logLevel: 'trace',
    logLevels: {
        webdriver: 'info',
        '@wdio/applitools-service': 'silent',
        '@wdio/devtools-service': 'silent'
    },
    bail: 5,
    baseUrl: process.env.baseUrl,
    waitforTimeout: 60000,
    connectionRetryTimeout: 120000,
    connectionRetryCount: 3,
    specFileRetriesDeferred: false,
    services: ['chromedriver', 'devtools'],
    framework: 'mocha',
    reporters: [
        'spec', 
        ['allure', {
            outputDir: 'allure-results',
            disableWebdriverStepsReporting: true,
            disableWebdriverScreenshotsReporting: false,
        }]
    ],
    mochaOpts: {
        ui: 'bdd',
        timeout: debug ? (24 * 60 * 60 * 1000) : defaultTimeout
    },
    beforeTest: function (test, context) {
        browser.maximizeWindow()
    },
    afterTest: function(test, context, { error, result, duration, passed, retries }) {
        if (!passed) {
            browser.takeScreenshot();
        }
        tests++;
        console.log('----------------------------------------------')
        console.log(`No of tests: ${tests}`)
        console.log(`error is:\n${error}`)
    }
}
```

**Describe the bug**
When any test fails with a timeOut error, code inside afterTest does not execute after that test. The hook does execute for tests that pass or fail with assertion errors.

**To Reproduce**
1. Implement some code inside afterTest hook
```
tests++;
console.log('----------------------------------------------')
console.log(`No of tests: ${tests}`)
```
2. Fail a test with timeOut error

**Expected behavior**
The code inside afterTest hook should be executed after the test fails with timeOut error

**Log**
https://gist.github.com/Harshad3009/fb0e465211ea5f48de92a483a157b315#file-logs-text

**Comment by christian-bromann** [comment]
Hey,
thanks for filing the issue. This seems indeed like bug. We would appreciate any help on this so please help us out! Have a look at our [contributing guidelines](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) and reach out if you have any questions. Cheers!

**Comment by christian-bromann** [comment]
This is a good first issue to pick for everyone. It seems that the `afterTest` hook is not executed. This is suppose to happen here: https://github.com/webdriverio/webdriverio/blob/master/packages/wdio-utils/src/test-framework/testFnWrapper.js#L81. Something in that function wrapper logic prevents the hook from being run.

To ensure we don't regress on this part of the PR should also be to add a smoke test that checks if `afterTest` hooks are being run even after a failure.

**Comment by unickq** [comment]
Subscribing

**Comment by worenga** [comment]
@christian-bromann the link you gave in 2019 is 404ing do you have any pointers where this issues can be debugged/fixed?

**Comment by christian-bromann** [comment]
I updated the link, see also https://webdriver.io/docs/contribute

**Comment by worenga** [comment]
Apologies @christian-bromann but it doesnt seem the link got updated?

**Comment by christian-bromann** [comment]
> but it doesnt seem the link got updated?

see also https://webdriver.io/docs/contribute

**Comment by ashishnarmen** [comment]
> Apologies @christian-bromann but it doesnt seem the link got updated?

The correct link is https://github.com/webdriverio/webdriverio/blob/master/packages/wdio-utils/src/test-framework/testFnWrapper.ts

**Comment by justusromijn** [comment]
I've been doing some investigation, but I just can't really pinpoint where it goes wrong. I've got a feeling it is something with the `switchSyncFlag` functionality, as I could follow up the code until there:
https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-utils/src/shim.ts#L427

**Comment by mhardin** [comment]
Just curious if there are any updates on this?  We are seeing the same issue in our test framework ('afterTest' hook gets skipped whenever a test times out).  This is causing issues for us because we set a lot of important metadata in our afterTest  hook, such as the video recording, etc.

**Comment by Sumit-Soman** [comment]
@christian-bromann can you please let us know by when can we expect fix for this issue. Or an alternate solution to handle it for time being. Need to take screenshot on failure with latest V7 Webdriverio Appium framework. Ant help here will be appreciated

**Comment by christian-bromann** [comment]
> can you please let us know by when can we expect fix for this issue.

@Sumit-Soman I think I made it very clear that help on this issue is very much appreciated. Don't wait until this gets fixed by the core team because this will likely not happen anytime soon. Get involved in the project and give back by contributing a fix for this.

**Comment by tjnk24** [comment]
There's something wrong with @wdio/local-runner. Downgraded to 7.12.1 version as its the latest version where the hook works correctly.

**Comment by LucianOutcomes** [comment]
Ran into this but only if the timeout occurred close to the end of the test. The webdriver session would close before the hook finished execution (in my case it is just a save screenshot function). Oddly enough I got it to behave by declaring the whole hook async 
```
async afterTest(test, context, { error, result, duration, passed, retries }) {
    if (!passed) {
      await browser.saveScreenshot(
        `${this.screenshotPath}/${test.title.replace(/ /g, '_').substring(0, 70)}` +
          `_` +
          `${(i += 1)}.png`,
      );
    }
  }
```
After a bit of reading up I'm not sure why this even does anything. Aren't the hooks async by default?

**Comment by christian-bromann** [comment]
> Aren't the hooks async by default?

Sure, but if you do an async operation in there like `saveScreenshot` you have to define it as `async` function.

**Comment by LucianOutcomes** [comment]
> > Aren't the hooks async by default?
> 
> Sure, but if you do an async operation in there like `saveScreenshot` you have to define it as `async` function.

Oh... makes sense. Maybe add a small snippet about this to the config file docu for dummies like me who thought that file was exempt from normal sync/async requirements ¯\_(ツ)_/¯

**Comment by BrLane** [comment]
We've written a custom reporter using WDIOReporter. When we encounter a test that times out, the test runner hangs. My first guess is that it's due to this bug. Could that be the case? If not I can dig around to see if this is a different known issue or open a bug if it's new.

**Comment by christian-bromann** [comment]
> Could that be the case? If not I can dig around to see if this is a different known issue or open a bug if it's new.

I don't think it is related as no one reported that the runner is hanging.

**Comment by BrLane** [comment]
> I don't think it is related as no one reported that the runner is hanging.

Ok. I'll look into our code to see if it's something we're doing. When I saw the hang behavior in our tests this bug was the first thing I thought about because the test failed due to timing out trying to find a (non-existent) element. FWIW, I verified the test times out and the runner continues executing without our custom reporter. Thanks!

**Comment by tamil777selvan** [comment]
Closing this as it's been fixed in #11172
